Adapt container settings when start

hqhq opened the pull request:
`adaptContainerSettings` is growing up, new it's only called
when create. It'll be a problem that old containers will never
have chance to adapt the latest rule. `HostConfig` of these
containers will be obsoleted.

Add this calling to start to avoid problems like #18550 and
avoid such backward compatability in the future.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

hqhq commented:
ping @runcom @estesp @cpuguy83 @calavera

thaJeztah commented:
and /cc @ibuildthecloud

ibuildthecloud commented on daemon/start.go:
```diff
@@ -36,9 +36,6 @@ func (daemon *Daemon) ContainerStart(name string, hostConfig *runconfig.HostConf
 				return err
 			}
 			container.Unlock()
-			if err := daemon.adaptContainerSettings(hostConfig, false); err != nil {
```
I would be concerned that the sequence of calls has changed with Windows.  No clue what windows is doing, but if for example `InitDNSHostConfig` needed some change this would cause an issue.

hqhq replied:
@ibuildthecloud This code path is for !windows, that doesn't matter, anyway I won't concern that because:
1. This path is for backword compatibility, which is deprecated for a very long time, so the usage is rare.
2. `adaptContainerSettings` in this path was added just very short ago, the impact is very limited.
3. The main change here is we call `adaptContainerSettings` later, adapt after verify is what we do for create, that should not be a problem.

What do you mean about "InitDNSHostConfig needed some change" and what issue would that cause?

estesp commented:
this needs a rebase, sorry @hqhq

hqhq committed:
Adapt container settings when start

`adaptContainerSettings` is growing up, new it's only called
when create. It'll be a problem that old containers will never
have chance to adapt the latest rule. `HostConfig` of these
containers will be obsoleted.

Add this calling to start to avoid problems like #18550 and
avoid such backward compatability in the future.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@estesp Rebased, thanks.

runcom commented:
LGTM

estesp commented:
LGTM
